File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/deeplink/DefaultDeepLinkInternalTest.java
package com.emarsys.mobileengage.deeplink;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;

import com.emarsys.core.api.result.CompletionListener;
import com.emarsys.core.device.DeviceInfo;
import com.emarsys.core.endpoint.ServiceEndpointProvider;
import com.emarsys.core.provider.timestamp.TimestampProvider;
import com.emarsys.core.provider.uuid.UUIDProvider;
import com.emarsys.core.request.RequestManager;
import com.emarsys.core.request.model.RequestModel;
import com.emarsys.core.storage.Storage;
import com.emarsys.mobileengage.MobileEngageRequestContext;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class DefaultDeepLinkInternalTest {

    private Activity mockActivity;
    private DeepLinkInternal deepLinkInternal;
    private RequestManager manager;
    private MobileEngageRequestContext requestContext;
    private TimestampProvider timestampProvider;
    private UUIDProvider uuidProvider;
    private DeviceInfo mockDeviceInfo;
    private ServiceEndpointProvider mockDeepLinkServiceProvider;
    private static final String APPLICATION_CODE = "applicationCode";
    private static final String DEEPLINK_BASE = "https://deep-link.eservice.emarsys.net/api/";
    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();


    @Before
    @SuppressWarnings("unchecked")
    public void init() {
        mockActivity = mock(Activity.class, Mockito.RETURNS_DEEP_STUBS);

        manager = mock(RequestManager.class);

        timestampProvider = mock(TimestampProvider.class);
        uuidProvider = mock(UUIDProvider.class);
        when(uuidProvider.provideId()).thenReturn("REQUEST_ID");
        mockDeviceInfo = mock(DeviceInfo.class);
        when(mockDeviceInfo.getSdkVersion()).thenReturn("0.0.1");
        requestContext = new MobileEngageRequestContext(
                APPLICATION_CODE,
                1,
                mockDeviceInfo,
                timestampProvider,
                uuidProvider,
                mock(Storage.class),
                mock(Storage.class),
                mock(Storage.class),
                mock(Storage.class)
        );

        mockDeepLinkServiceProvider = mock(ServiceEndpointProvider.class);
        when(mockDeepLinkServiceProvider.provideEndpointHost()).thenReturn(DEEPLINK_BASE);
        deepLinkInternal = new DefaultDeepLinkInternal(manager, requestContext, mockDeepLinkServiceProvider);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_requestManagerMustNotBeNull() {
        new DefaultDeepLinkInternal(null, requestContext, mockDeepLinkServiceProvider);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_requestContextMustNotBeNull() {
        new DefaultDeepLinkInternal(manager, null, mockDeepLinkServiceProvider);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_deepLinkServiceProvider_mustNotBeNull() {
        new DefaultDeepLinkInternal(manager, requestContext, null);
    }

    @Test
    public void testTrackDeepLink_doesNotCrashOnNonHierarchicalUris() {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("mailto:a@b.com"));

        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, null);
    }

    @Test
    public void testTrackDeepLink_requestManagerCalled_withCorrectRequestModel() {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&ems_dl=1_2_3_4_5"));

        Map<String, Object> payload = new HashMap<>();
        payload.put("ems_dl", "1_2_3_4_5");

        Map<String, String> headers = new HashMap<>();
        headers.put("User-Agent",
                String.format("Emarsys SDK %s Android %s", requestContext.getDeviceInfo().getSdkVersion(), Build.VERSION.SDK_INT));

        RequestModel expected = new RequestModel.Builder(timestampProvider, uuidProvider)
                .url(DEEPLINK_BASE + "clicks")
                .headers(headers)
                .payload(payload)
                .build();

        ArgumentCaptor<RequestModel> captor = ArgumentCaptor.forClass(RequestModel.class);

        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, null);

        verify(manager).submit(captor.capture(), (CompletionListener) isNull());

        RequestModel result = captor.getValue();
        assertRequestModels(expected, result);
    }

    @Test
    public void testTrackDeepLink_requestManagerCalled_withCorrectCompletionHandler() {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&ems_dl=1_2_3_4_5"));

        CompletionListener completionListener = mock(CompletionListener.class);

        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, completionListener);

        verify(manager).submit(any(RequestModel.class), eq(completionListener));
    }

    @Test
    public void testTrackDeepLink_setsClickedFlag_onIntentBundle() {
        Intent originalIntent = mock(Intent.class);
        when(mockActivity.getIntent()).thenReturn(originalIntent);

        Intent currentIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&ems_dl=1_2_3_4_5"));

        deepLinkInternal.trackDeepLinkOpen(mockActivity, currentIntent, null);

        verify(originalIntent).putExtra("ems_deep_link_tracked", true);
    }

    @Test
    public void testTrackDeepLink_doesNotCallRequestManager_whenTrackedFlagIsSet() {
        Intent intentFromActivity = mock(Intent.class);
        when(mockActivity.getIntent()).thenReturn(intentFromActivity);
        when(intentFromActivity.getBooleanExtra("ems_deep_link_tracked", false)).thenReturn(true);

        Intent currentIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&ems_dl=1_2_3_4_5"));

        deepLinkInternal.trackDeepLinkOpen(mockActivity, currentIntent, null);

        verify(manager, times(0)).submit(any(RequestModel.class), (CompletionListener) isNull());
    }

    @Test
    public void testTrackDeepLink_doesNotCallRequestManager_whenDataIsNull() {
        Intent intent = new Intent();

        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, null);

        verify(manager, times(0)).submit(any(RequestModel.class), (CompletionListener) isNull());
    }

    @Test
    public void testTrackDeepLink_doesNotCallRequestManager_whenUriDoesNotContainEmsDlQueryParameter() {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&other=1_2_3_4_5"));

        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, null);

        verify(manager, times(0)).submit(any(RequestModel.class), (CompletionListener) isNull());
    }

    private void assertRequestModels(RequestModel expected, RequestModel result) {
        assertEquals(expected.getUrl(), result.getUrl());
        assertEquals(expected.getMethod(), result.getMethod());
        assertEquals(expected.getPayload(), result.getPayload());
        assertEquals(expected.getHeaders(), result.getHeaders());
    }
}

File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/deeplink/DefaultDeepLinkInternalTest.kt
package com.emarsys.mobileengage.deeplink

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Build
import com.emarsys.core.api.result.CompletionListener
import com.emarsys.core.device.DeviceInfo
import com.emarsys.core.endpoint.ServiceEndpointProvider
import com.emarsys.core.provider.timestamp.TimestampProvider
import com.emarsys.core.provider.uuid.UUIDProvider
import com.emarsys.core.request.RequestManager
import com.emarsys.core.request.model.RequestModel
import com.emarsys.mobileengage.MobileEngageRequestContext
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.junit.Before
import org.junit.Test

class DefaultDeepLinkInternalTest  {
    private lateinit var mockActivity: Activity
    private lateinit var deepLinkInternal: DeepLinkInternal
    private lateinit var mockManager: RequestManager
    private lateinit var requestContext: MobileEngageRequestContext
    private lateinit var mockTimestampProvider: TimestampProvider
    private lateinit var mockUuidProvider: UUIDProvider
    private lateinit var mockDeviceInfo: DeviceInfo
    private lateinit var mockDeepLinkServiceProvider: ServiceEndpointProvider

    @Before
    fun init() {
        mockActivity = mockk(relaxed = true)
        mockManager = mockk(relaxed = true)
        mockTimestampProvider = mockk(relaxed = true)
        mockUuidProvider = mockk(relaxed = true)
        every { mockUuidProvider.provideId() } returns "REQUEST_ID"
        mockDeviceInfo = mockk(relaxed = true)
        every { mockDeviceInfo.sdkVersion } returns "0.0.1"
        requestContext = MobileEngageRequestContext(
            APPLICATION_CODE,
            1,
            null,
            mockDeviceInfo,
            mockTimestampProvider,
            mockUuidProvider,
            mockk(relaxed = true),
            mockk(relaxed = true),
            mockk(relaxed = true),
            mockk(relaxed = true),
            mockk(relaxed = true),
            mockk(relaxed = true)
        )

        mockDeepLinkServiceProvider = mockk(relaxed = true)
        every { mockDeepLinkServiceProvider.provideEndpointHost() } returns DEEPLINK_BASE
        deepLinkInternal =
            DefaultDeepLinkInternal(requestContext, mockDeepLinkServiceProvider, mockManager)
    }

    @Test
    fun testTrackDeepLink_doesNotCrashOnNonHierarchicalUris() {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse("mailto:a@b.com"))
        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, null)
    }

    @Test
    fun testTrackDeepLink_requestManagerCalled_withCorrectRequestModel() {
        val intent = Intent(
            Intent.ACTION_VIEW,
            Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&ems_dl=1_2_3_4_5")
        )
        val payload: MutableMap<String, Any?> = HashMap()
        payload["ems_dl"] = "1_2_3_4_5"
        val headers: MutableMap<String, String> = HashMap()
        headers["User-Agent"] = String.format(
            "Emarsys SDK %s Android %s",
            requestContext.deviceInfo.sdkVersion,
            Build.VERSION.SDK_INT
        )
        val expected = RequestModel.Builder(mockTimestampProvider, mockUuidProvider)
            .url("$DEEPLINK_BASE/api/clicks")
            .headers(headers)
            .payload(payload)
            .build()

        val requestModelSlot = slot<RequestModel>()

        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, null)

        verify { mockManager.submit(capture(requestModelSlot), null) }
        assertRequestModels(expected, requestModelSlot.captured)
    }

    @Test
    fun testTrackDeepLink_requestManagerCalled_withCorrectCompletionHandler() {
        val completionListener: CompletionListener = mockk(relaxed = true)
        val intent = Intent(
            Intent.ACTION_VIEW,
            Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&ems_dl=1_2_3_4_5")
        )

        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, completionListener)
        verify { mockManager.submit(any(), completionListener) }
    }

    @Test
    fun testTrackDeepLink_setsClickedFlag_onIntentBundle() {
        val originalIntent: Intent = mockk(relaxed = true)
        every { mockActivity.intent } returns originalIntent
        val currentIntent = Intent(
            Intent.ACTION_VIEW,
            Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&ems_dl=1_2_3_4_5")
        )
        deepLinkInternal.trackDeepLinkOpen(mockActivity, currentIntent, null)
        verify { originalIntent.putExtra("ems_deep_link_tracked", true) }
    }

    @Test
    fun testTrackDeepLink_doesNotCallRequestManager_whenTrackedFlagIsSet() {
        val intentFromActivity: Intent = mockk(relaxed = true)
        every { mockActivity.intent } returns intentFromActivity
        every { intentFromActivity.getBooleanExtra("ems_deep_link_tracked", false) } returns true
        val currentIntent = Intent(
            Intent.ACTION_VIEW,
            Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&ems_dl=1_2_3_4_5")
        )
        deepLinkInternal.trackDeepLinkOpen(mockActivity, currentIntent, null)
        verify(exactly = 0) { mockManager.submit(any(), any()) }
    }

    @Test
    fun testTrackDeepLink_doesNotCallRequestManager_whenDataIsNull() {
        val intent = Intent()
        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, null)
        verify(exactly = 0) { mockManager.submit(any(), any()) }
    }

    @Test
    fun testTrackDeepLink_doesNotCallRequestManager_whenUriDoesNotContainEmsDlQueryParameter() {
        val intent = Intent(
            Intent.ACTION_VIEW,
            Uri.parse("https://demo-mobileengage.emarsys.net/something?fancy_url=1&other=1_2_3_4_5")
        )
        deepLinkInternal.trackDeepLinkOpen(mockActivity, intent, null)
        verify(exactly = 0) { mockManager.submit(any(), any()) }
    }

    private fun assertRequestModels(expected: RequestModel, result: RequestModel) {
        result.url shouldBe expected.url
        result.method shouldBe expected.method
        result.payload shouldBe expected.payload
        result.headers shouldBe expected.headers
    }

    companion object {
        private const val APPLICATION_CODE = "applicationCode"
        private const val DEEPLINK_BASE = "https://deep-link.eservice.emarsys.net"
    }
}
